game/debriefing.py

```diff
@@ -109,16 +109,43 @@ class StateData:
     base_capture_events: List[str]
 
     @classmethod
-    def from_json(cls, data: Dict[str, Any]) -> StateData:
+    def from_json(cls, data: Dict[str, Any], unit_map: UnitMap) -> StateData:
+        def clean_unit_list(unit_list: List[Any]) -> List[str]:
+            # Cleans list of units in state.json by
+            # - Removing duplicates. Airfields emit a new "dead" event every time a bomb
+            #   is dropped on them when they've already dead.
+            # - Normalise dead map objects (which are ints) to strings. The unit map
+            #   only stores strings
+            units = set()
+            for unit in unit_list:
+                units.add(str(unit))
+            return list(units)
+
+        killed_aircraft = []
+        killed_ground_units = []
+
+        # process killed units from S_EVENT_UNIT_LOST & S_EVENT_KILL
+        killed_units = clean_unit_list(data["unit_lost_events"] + data["kill_events"])
+        for unit in killed_units:  # organize killed units into aircraft vs ground
+            if unit_map.flight(unit) is not None:
+                killed_aircraft.append(unit)
+            else:
+                killed_ground_units.append(unit)
+
+        # process killed units from S_EVENT_CRASH
+        for unit in data["crash_events"]:  # no need to clean crash events
```

Comment:
> no need to clean crash events

Why not? DCS presumably won't emit a duplicate crash event, but don't you need to dedup everything on L147-148? DCS might not emit a duplicate crash, but might it emit a kill, crash, dead, and unit lost all for one unit?

If I'm right about that, I think what you're better off making `killed_aircraft` and `killed_ground_units` into `set[str]` and just doing the list conversion at the end?

If not, could you expand the comment explaining why that doesn't matter?

Reply:
Good point, we do need to deduplicate all event types. Have updated the code to do this.